packages/muelu/src/Misc/MueLu_AggregateQualityEstimateFactory_decl.hpp

```diff
@@ -0,0 +1,155 @@
+#define MUELU_AGGREGATEQUALITYESTIMATEFACTORY_DECL_HPP
+
+#include "MueLu_ConfigDefs.hpp"
+#include "MueLu_SingleLevelFactoryBase.hpp"
+#include "MueLu_AggregateQualityEstimateFactory_fwd.hpp"
+
+#include <Xpetra_CrsMatrix_fwd.hpp>
+#include <Xpetra_Matrix_fwd.hpp>
+#include <Xpetra_MultiVector_fwd.hpp>
+#include <Xpetra_MapFactory_fwd.hpp>
+#include <Xpetra_Map_fwd.hpp>
+#include <Xpetra_Matrix_fwd.hpp>
+#include <Xpetra_MultiVector_fwd.hpp>
+#include <Xpetra_MultiVectorFactory_fwd.hpp>
+#include <Xpetra_Import_fwd.hpp>
+#include <Xpetra_ImportFactory_fwd.hpp>
+#include <Xpetra_CrsMatrixWrap_fwd.hpp>
+
+#include "MueLu_Aggregates_fwd.hpp"
+#include "MueLu_Level_fwd.hpp"
+
+namespace MueLu {
+
+  /*!
+    @class AggregateQualityEstimateFactory class.
+    @brief An factory which assigns each aggregate a quality
+           estimate. Originally developed by Napov and Notay in the
+	   context of plain aggregation, while this quality estimate
+	   does not correspond to a robust convergence guarentee (as
+	   it does for plain aggregation), we find empirically that
+	   it is a good way of discovering poorly constructed aggregates
+	   even in the smoothed aggregation context.
+
+	   Napov, A., & Notay, Y. (2012). An algebraic multigrid method
+	   with guaranteed convergence rate. SIAM journal on scientific
+	   computing, 34(2), A1079-A1109.
+  */
+
+  template <class Scalar = double, class LocalOrdinal = int, class GlobalOrdinal = LocalOrdinal, class Node = KokkosClassic::DefaultNode::DefaultNodeType>
+  class AggregateQualityEstimateFactory : public SingleLevelFactoryBase {
+#undef MUELU_AGGREGATEQUALITYESTIMATEFACTORY_SHORT
+#include "MueLu_UseShortNames.hpp"
+
+  public:
+      //! @name Constructors/Destructors.
+      //@{
+
+      //! Constructor.
+      AggregateQualityEstimateFactory();
+
+      //! Destructor.
+      virtual ~AggregateQualityEstimateFactory();
+
+      //@}
+
+      RCP<const ParameterList> GetValidParameterList() const;
+
+      //! @name Input
+      //@{
+
+      /*! @brief Specifies the data that this class needs, and the factories that generate that data.
+
+        If the Build method of this class requires some data, but the generating factory is not specified in DeclareInput, then this class
+        will fall back to the settings in FactoryManager.
+      */
+      void DeclareInput(Level &currentLevel) const;
+
+      //@}
+
+      //! @name Build methods.
+      //@{
+
+      //! Build aggregate quality esimates with this factory.
+      void Build(Level & currentLevel) const;
+      void BuildP(Level & currentLevel) const;
+
+      //@}
+
+      //! @name Utility method to convert aggregate data to a convenient format.
+      //@{
+
+      //! Build aggregate quality esimates with this factory.
+      static void ConvertAggregatesData(RCP<const Aggregates> aggs, Teuchos::ArrayRCP<LO>& aggSortedVertices, Teuchos::ArrayRCP<LO>& aggsToIndices, Teuchos::ArrayRCP<LO>& aggSizes);
```

Comment:
More of a comment than a review: `Teuchos::Array`, `Teuchos::ArrayView`, `Teuchos::ArrayRCP` and `Teuchos::RCP` and a few other related and useful classes are imported in the 'muelu' namespace, so no need to add the `Teuchos` prefix. See this file: `muelu/src/Headers/MueLu_ConfigDefs.hpp`